### ISSUE TITLE
Vite seems to be bundling dependencies it shouldn't?

### ISSUE THREAD

**Issue by Nickersoft** [withastro/astro#11785]
### Astro Info

```block
Astro                    v4.14.2
Node                     v20.16.0
System                   macOS (arm64)
Package Manager          pnpm
Output                   hybrid
Adapter                  @astrojs/node
Integrations             @astrojs/react
                         @astrojs/tailwind
```

### Describe the Bug

I apologize in advance that I don't have a _minimal_ reproduction of this, as I'm not sure what is actually causing the behavior to begin with (I do have a repro though).

I'm currently trying to build my Astro project inside a [PNPM monorepo](https://github.com/Nickersoft/code-snippets/tree/nickersoft/build-bug/www) and am referencing a shared private monorepo package. 

I'm only using the methods from it to prerender a few pages and JSON routes, however, Astro/Vite seems to be trying to bundle all of the code into `dist/server/chunks`, including some CJS dependencies that rely on `__filename` and result in the following error when building:

```
__filename is not defined in ES module scope
```

I don't have a firm grasp on how Astro actually prerenders routes, so I'm not sure what I can change to try to fix this. I originally [opened an issue](https://github.com/vitejs/vite/issues/17902) with the Vite team but they redirected me back here seeing Astro has a lot of moving parts and maybe the way the compiler works is causing this.

To reproduce, you just run `pnpm install` in my linked repo then run `pnpm build` from inside the `www` directory.

### What's the expected result?

Expected to be able to use a local monorepo package and have it prerender routes without issue (or at lest transform the CJS).

### Link to Minimal Reproducible Example

https://github.com/Nickersoft/code-snippets/tree/nickersoft/build-bug/www

### Participation

- [ ] I am willing to submit a pull request for this issue.

**Comment by ematipico** [comment]
The person from vite already did a brief analysis, which I advise you to follow.

You have some development dependencies inside `dependecies`. Move them in `devDependecies`

**Comment by Nickersoft** [comment]
Hey @ematipico,

I appreciate the prompt response – to answer your first question, I believe it is expected? Otherwise I'd have to have deep path imports for every method I'd like to call, so I use an `index.ts` to export the methods I need in my Astro app and related projects.

I'm unfortunately in a bit of a tight spot, because the dependencies that were mentioned in the other thread – ones that _are_ using dev dependencies like Prettier and TypeScript – I'm using to format and modify TypeScript ASTs on the server (prerendered JSON routes). So I would think they _should_ be regular dependencies, unless they're dev in the sense that they're only required to prerender server routes during build-time?

**Comment by Nickersoft** [comment]
So I did a bit more digging here and it looks like the error is because Vite is bundling `ts-morph`, which is a dependency of my shared library `registry-tools`. Adding `ts-morph` as a direct dependency of my Astro project allows it to build fine. 

I sort of understand _why_ this is happening, seeing `registry-tools` is a private library and its code is embedded in the Astro scripts (and therefore when it references `ts-morph` Astro doesn't know where to find it because it's not a direct dependency, so it bundles it too).

Apart from adding `ts-morph` to my dependencies, is there really any other kind of fix? I tried making `ts-morph` a _peer_ dependency of my shared library instead, but it still errors. Just a bit curious as to why this happens and other strategies for avoiding it. Interestingly, `ts-morph` (a CJS-only package) is the only dependency this is happening for.
